### PR TITLE
feature(zt): add curiosity icm algorithm

### PR DESCRIPTION
**Pull request by timothijoe**
Inverse Dynamic Features implementation based on the Paper "Curiosity-driven Exploration by Self-supervised Prediction"

### CODE REVIEW

File: ding/reward_model/curiosity_reward_model.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import torch.nn as nn
+import torch.optim as optim
+import torch.nn.functional as F
+
+from ding.utils import SequenceType, REWARD_MODEL_REGISTRY
+from ding.model import FCEncoder, ConvEncoder
+from .base_reward_model import BaseRewardModel
+
+
+def collect_states(iterator):
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:15]: add python typing lint

File: ding/reward_model/curiosity_reward_model.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        if isinstance(obs_shape, int) or len(obs_shape) == 1:
+            self.feature= FCEncoder(obs_shape, hidden_size_list)
+        elif len(obs_shape) == 3:
+            self.feature= ConvEncoder(obs_shape, hidden_size_list)
+        else:
+            raise KeyError(
+                "not support obs_shape for pre-defined encoder: {}, please customize your own RND model".
+                format(obs_shape)
+            )
+        self.device = 'cpu' #torch.device('cuda' if use_cuda else 'cpu')
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:41]: Don't hardcode device here, pass `device` as the arguments of constructor method.

File: ding/reward_model/curiosity_reward_model.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    def __init__(self, obs_shape: Union[int, SequenceType], hidden_size_list: SequenceType, action_shape:int) -> None:
+        super(CuriosityNetwork, self).__init__()
+        if isinstance(obs_shape, int) or len(obs_shape) == 1:
+            self.feature= FCEncoder(obs_shape, hidden_size_list)
+        elif len(obs_shape) == 3:
+            self.feature= ConvEncoder(obs_shape, hidden_size_list)
+        else:
+            raise KeyError(
+                "not support obs_shape for pre-defined encoder: {}, please customize your own RND model".
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:38]: Modify corresponding statements like `RND model`

File: ding/reward_model/curiosity_reward_model.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            )
+        self.device = 'cpu' #torch.device('cuda' if use_cuda else 'cpu')
+        self.action_shape = action_shape
+        feature_output = hidden_size_list[-1]
+        self.inverse_net = nn.Sequential(
+            nn.Linear(feature_output * 2, 512),
+            nn.ReLU(),
+            nn.Linear(512, action_shape)
+        )
+        self.residual = [nn.Sequential(
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:49]: Don't use `[object()] * 8` but `[object() for _ in range(8)]`, the former will share the same object with different reference

File: ding/reward_model/curiosity_reward_model.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        train_data_list = [i for i in range(0, len(self.train_states))]
+        train_data_index = random.sample(train_data_list, self.cfg.batch_size)
+        data_states = [self.train_states[i] for i in train_data_index]
+        data_states = torch.stack(data_states)
+        data_next_states = [self.train_next_states[i] for i in train_data_index]
+        data_next_states = torch.stack(data_next_states)
+        data_actions = [self.train_actions[i] for i in train_data_index]
+        data_actions = torch.stack(data_actions)
+        real_next_state_feature, pred_next_state_feature, pred_action = self.reward_model(data_states, data_next_states, data_actions)
+        ce = nn.CrossEntropyLoss(reduction="mean")
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:139]: Loss module should create in `__init__` method

File: ding/reward_model/curiosity_reward_model.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        forward_mse = nn.MSELoss()
+        inverse_loss = ce(pred_action, data_actions.squeeze(dim=1).long())
+        forward_loss = forward_mse(pred_next_state_feature, real_next_state_feature.detach())
+        #loss = F.mse_loss(predict_feature, target_feature.detach())
+        loss = inverse_loss + forward_loss
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:144]: It would be better to add a factor of balancing two kinds of loss

File: ding/reward_model/curiosity_reward_model.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            nn.Linear(action_shape + feature_output, 512),
+            nn.LeakyReLU()
+        )
+        self.forward_net_2 = nn.Sequential(
+            nn.Linear(action_shape + 512, feature_output),
+        )
+
+    def forward(self, state, next_state, action_long):
+        #state, next_state, action = inputs
+        state = torch.FloatTensor(state)
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:65]: You need to consider cuda device cases whenever you create tensor.

File: ding/reward_model/curiosity_reward_model.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        pred_action = torch.cat((encode_state, encode_next_state), 1)
+        pred_action = self.inverse_net(pred_action)
+        # ---------------------
+
+        # get pred next state
+        pred_next_state_feature_orig = torch.cat((encode_state, action), 1)
+        pred_next_state_feature_orig = self.forward_net_1(pred_next_state_feature_orig)
+
+        # residual
+        for i in range(4):
```
Comment (puyuan1996) [inline ding/reward_model/curiosity_reward_model.py:81]: 可以在minigrid_empty8上测试一下，前向模型不包括residual网络部分会对结果有什么样的影响

File: ding/reward_model/curiosity_reward_model.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        )
+        self.forward_net_2 = nn.Sequential(
+            nn.Linear(action_shape + 512, feature_output),
+        )
+
+    def forward(self, state: torch.Tensor, next_state: torch.Tensor, action_long: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        state = torch.FloatTensor(state)
+        next_state = torch.FloatTensor(next_state)
+        action_long = torch.LongTensor(action_long)
+        action_onehot = torch.FloatTensor(len(action_long), self.action_shape)
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:65]: you can get one_hot result using function——`from ding.torch_utils import one_hot`

File: ding/reward_model/curiosity_reward_model.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+        self.intrinsic_reward_type = config.intrinsic_reward_type
+        assert self.intrinsic_reward_type in ['add', 'new', 'assign']
+        self.train_data = []
+        self.train_states = []
+        self.train_next_states = []
+        self.train_actions = [] 
+        self.opt = optim.Adam(self.reward_model.parameters(), config.learning_rate)
+        self.ce = nn.CrossEntropyLoss(reduction="mean")
+        self.forward_mse = nn.MSELoss()
+        self.reverse_scale = 1
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:122]: It would be better if this argument can be accessed by `self.cfg`, and you can add default value in `config`

File: ding/reward_model/curiosity_reward_model.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        action_long = torch.LongTensor(action_long)
+        action_onehot = torch.FloatTensor(len(action_long), self.action_shape)
+        action_onehot.zero_()
+        action_onehot.scatter_(1, action_long.view(len(action_long), -1), 1)
+        action = action_onehot
+        encode_state = self.feature(state)
+        encode_next_state = self.feature(next_state)
+        # get pred action
+        pred_action = torch.cat((encode_state, encode_next_state), 1)
+        pred_action = self.inverse_net(pred_action)
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:73]: This variable should be named as `logit` action, it is not `action` but `logit`:
- `action = torch.multinomial(torch.softmax(logit, dim=1), 1)`

File: ding/reward_model/curiosity_reward_model.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        ) for _ in range(8)]
+        self.forward_net_1 = nn.Sequential(
+            nn.Linear(action_shape + feature_output, 512),
+            nn.LeakyReLU()
+        )
+        self.forward_net_2 = nn.Sequential(
+            nn.Linear(action_shape + 512, feature_output),
+        )
+
+    def forward(self, state: torch.Tensor, next_state: torch.Tensor, action_long: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:61]: Please add shaps information for neural network implementation, like [this](https://github.com/opendilab/DI-engine/blob/main/ding/model/template/q_learning.py#L86)

File: ding/reward_model/curiosity_reward_model.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+            pred_next_state_feature = self.residual[i * 2](torch.cat((pred_next_state_feature_orig, action), 1))
+            pred_next_state_feature_orig = self.residual[i * 2 + 1](
+                torch.cat((pred_next_state_feature, action), 1)) + pred_next_state_feature_orig
+        pred_next_state_feature = self.forward_net_2(torch.cat((pred_next_state_feature_orig, action), 1))
+        real_next_state_feature = encode_next_state
+        return real_next_state_feature, pred_next_state_feature, pred_action_logit
+
+
+@REWARD_MODEL_REGISTRY.register('curiosity')
+class ICMRewardModel(BaseRewardModel):
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:108]: You can add original paper reference in class overview comment like other algorithms

Comment (timothijoe) [reply ding/reward_model/curiosity_reward_model.py:115]: I have added the original paper reference on the ICMNetwork Description

File: dizoo/minigrid/config/minigrid_ppo_curiosity_config.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
+from easydict import EasyDict
```
Comment (PaParaZz1) [inline dizoo/minigrid/config/minigrid_ppo_curiosity_config.py:1]: add pong icm config

File: dizoo/minigrid/config/minigrid_ppo_curiosity_config.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+from ding.entry import serial_pipeline_reward_model
+
+minigrid_ppo_rnd_config = dict(
+    exp_name='minigrid_empty8_ppo_rnd',  #exp_name='minigrid_empty8_ppo_rnd',
```
Comment (PaParaZz1) [inline dizoo/minigrid/config/minigrid_ppo_curiosity_config.py:5]: modify icm exp name

File: ding/reward_model/curiosity_reward_model.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+            pred_next_state_feature_orig = self.residual[i * 2 + 1](
+                torch.cat((pred_next_state_feature, action), 1)
+            ) + pred_next_state_feature_orig
+        pred_next_state_feature = self.forward_net_2(torch.cat((pred_next_state_feature_orig, action), 1))
+        real_next_state_feature = encode_next_state
+        return real_next_state_feature, pred_next_state_feature, pred_action_logit
+
+
+@REWARD_MODEL_REGISTRY.register('curiosity')
+class ICMRewardModel(BaseRewardModel):
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:115]: add comments like this [file](https://github.com/opendilab/DI-engine/blob/main/ding/reward_model/gail_irl_model.py#L108) and config comments like [this](https://github.com/opendilab/DI-engine/blob/main/ding/policy/acer.py#L57)

File: ding/reward_model/curiosity_reward_model.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        self.reward_model.to(self.device)
+        self.intrinsic_reward_type = config.intrinsic_reward_type
+        assert self.intrinsic_reward_type in ['add', 'new', 'assign']
+        self.train_data = []
+        self.train_states = []
+        self.train_next_states = []
+        self.train_actions = []
+        self.opt = optim.Adam(self.reward_model.parameters(), config.learning_rate)
+        self.ce = nn.CrossEntropyLoss(reduction="mean")
+        self.forward_mse = nn.MSELoss(reduction='none')
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:144]: why use `reduction='none'`, `reduction='mean'` is enough here, you don't need to do mean operation in the following related codes

File: ding/reward_model/curiosity_reward_model.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+            - action_long (:obj:`torch.Tensor`):
+                The action, (B,M=action_size)``.
+        Returns:
+            - real_next_state_feature (:obj:`torch.Tensor`):
+                Run with the encoder. Return the real next_state's embedded feature.
+            - pred_next_state_feature (:obj:`torch.Tensor`):
+                Run with the encoder and residual network. Return the predicted next_state's embedded feature.
+            - pred_action_logit (:obj:`Dict`):
+                Run with the encoder. Return the predicted action logit.
+        """
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:87]: add shapes comments for this method, like [this](https://github.com/opendilab/DI-engine/blob/main/ding/model/template/q_learning.py#L76)

File: ding/reward_model/curiosity_reward_model.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+        self.inverse_net = nn.Sequential(nn.Linear(feature_output * 2, 512), nn.ReLU(), nn.Linear(512, action_shape))
+        self.residual = [
+            nn.Sequential(
+                nn.Linear(action_shape + 512, 512),
+                nn.LeakyReLU(),
+                nn.Linear(512, 512),
+            ) for _ in range(8)
+        ]
+        self.forward_net_1 = nn.Sequential(nn.Linear(action_shape + feature_output, 512), nn.LeakyReLU())
+        self.forward_net_2 = nn.Sequential(nn.Linear(action_shape + 512, feature_output), )
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:65]: you can just use `self.forward_net_2 = nn.Linear(action_shape + 512, feature_output)` here

File: ding/reward_model/curiosity_reward_model.py [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+                Run with the encoder. Return the predicted action logit.
+        """
+        state = torch.FloatTensor(state)
+        next_state = torch.FloatTensor(next_state)
+        action_long = torch.LongTensor(action_long)
+        action = one_hot(action_long, num=self.action_shape).squeeze(1)
+        encode_state = self.feature(state)
+        encode_next_state = self.feature(next_state)
+        # get pred action logit
+        pred_action_logit = torch.cat((encode_state, encode_next_state), 1)
```
Comment (PaParaZz1) [inline ding/reward_model/curiosity_reward_model.py:95]: change this name `pred_action_logit` here to `concat_state`

File: ding/policy/ppo.py [hunk 21]
```diff
@@ -632,6 +632,7 @@ def _process_transition(self, obs: Any, model_output: dict, timestep: namedtuple
         if not self._nstep_return:
```
Comment (PaParaZz1) [inline ding/policy/ppo.py:632]: use the same implementation in both cases here

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Comment by timothijoe** [comment]
ICM_doorkey8x8
A picture of doorkey 8x8 with three seeds ( 0, 10, 20).  All three seeds converges to 0.96 finally

**Review by PaParaZz1 — APPROVED** [review]
